### DISCUSSION TITLE
Should router abort connection if no_such_registration or no_such_subscription?

### DISCUSSION THREAD

**Discussion by SunFulong** [wamp-proto/wamp-proto#534] [Q&A]
Last time, we discussed if a callee yield a non-existed invocation.request, router should abort the connection.

So, what if a callee unregister a non-existed register id? Should router abort it too?

**Comment by Jopie64** [comment]
I would be ok if it did because this cannot have races. Unless... a registration can be forcefully unregistered via the router meta API...
Can it? 

If so, then this should be explicitly mentioned. A router should not abort a connection in this case unless it fixed the race somehow.

**Comment by SunFulong** [comment]
> a registration can be forcefully unregistered via the router meta API...
> Can it?

Check again, didn't found force unregister or unsubscribe procedure in meta APIs.

**Comment by oberstet** [comment]
> didn't found force unregister

see here in Crossbar.io:

https://github.com/crossbario/crossbar/blob/0089c1ef6fbbb87fc7316088a91f1859fa84eeb0/crossbar/router/service.py#L508

**Comment by oberstet** [comment]
> unregister a non-existed register id

see here in Crossbar.io:

https://github.com/crossbario/crossbar/blob/0089c1ef6fbbb87fc7316088a91f1859fa84eeb0/crossbar/router/dealer.py#L597

**Comment by oberstet** [comment]
@SunFulong btw, discussing and comparing router behavior would be easier by directly comparing/linking code: is your Rust based WAMP router code - that you talked about before - published and accessible by now/already?

**Reply by SunFulong** [reply]
> is your Rust based WAMP router code published and accessible by now/already?

Not yet, still in progress, only implemented RPC part, you can feel my progress via my issues, PRs and discusses.

**Reply by oberstet** [reply]
Thanks for your response! But, hmm ...

> Not yet, still in progress

Sure, no problem! Release early, release often. You should release / publish source before you have even finished your first line of code. Please don't get me wrong, I am walking my talk: that's how both Autobahn and Crossbar.io have been developed.

> you can feel my progress via my issues, PRs and discusses.

Sure. This is not how OSS development is supposed to work. You unnecessarily make "helping" much harder. I don't care "feeling" your progress;)

Taking this further, I am not only much less inclined to spend unpaid free time to "help", but I am also wondering as a consequence: what is your actual agenda?

Sorry, but rgd topics of OSS, security, decentralization and right-of-privacy & anonymity, I am pretty .. some would say "opinionated", I say "logical" and "consistent" ;)

**Reply by SunFulong** [reply]
> You should release / publish source before you have even finished your first line of code.

My project starting point is wrong, although it can run (in my production environment), but really unstable, and all comments are Chinese not English, so I don't think publish it now is good choice.

I will publish it when change to right way, and at least fully implemented basic profile.

> I am not only much less inclined to spend unpaid free time to "help"

I'm sorry for taking up your time, but the questions I asked are indeed what I encountered and confused during the development process, not just random ideas.

> but I am also wondering as a consequence: what is your actual agenda?

Actual agenda? What agenda can I have? Just build and run, when stable (in my eyes), publish to public, that all.

**Reply by oberstet** [reply]
thanks for explaining your thinking. it's all obviously your decision whether at all, and when you publish your code!

**Comment by oberstet** [comment]
> Last time, we discussed if a callee yield a non-existed invocation.request, router should abort the connection.

in my eyes, it should abort the callee session, **BUT only if we added the ability to distinguish between a bogus invocation request ID and a race condition** (as noted / brought up by @Jopie64), and if it indeed wasn't a race but inherently bogus

in the absence of such means to differentiate between those 2 cases, ignoring seems best

btw, this is also what Crossbar.io currently does:

https://github.com/crossbario/crossbar/blob/0089c1ef6fbbb87fc7316088a91f1859fa84eeb0/crossbar/router/dealer.py#L1413

> So, what if a callee unregister a non-existed register id? Should router abort it too?

nope, there is a regular error for that

https://github.com/crossbario/autobahn-python/blob/f38f16ba28fa253dee951068cc729089c88d857d/autobahn/wamp/exception.py#L135

in case of "unregister", the router is expected to send a reply regularily, and that reply might then be above error

in case of "yield", there is no router reply *to the callee* (https://wamp-proto.org/wamp_bp_latest_ietf.html#name-calling-and-invocations), and hence no way to transport an error

why is using a bogus/invalid ID in `YIELD` different from using a bogus/invalid ID in `UNREGISTER`?

good question;)

I would argue, a YIELD is always in the context of a *single* WAMP *action* in a session (the call being handled), whereas an UNREGISTER like REGISTER is about deestablishing / establishing a new *state* (the procedure registration / registered procedure)

following this argument, we might want to *distinguish handling of IDs received by a peer that cannot be processed by that peer depending on ID scope*:

* global & router scope
* session scope

https://wamp-proto.org/wamp_bp_latest_ietf.html#section-2.1.2
